### DISCUSSION TITLE
Solving steady-state problem using explicit solver?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17368] [Q&A General] [answered]
Hi all,

Is there a way to solve a steady-state problem using explicit method?

I know MOOSE is optimised with implicit, but my problem is kinda large (couple million elements) and require to run many steady-state simulations (with a minor change in material property at each run), so my supervisor suggests me to work with explicit solver instead.

Ive attached a simplify version the input file (tensor mech with implicit) here.
[implicit_to_explicit.zip](https://github.com/idaholab/moose/files/6160434/implicit_to_explicit.zip)
Thank you guys.

Kind regards,
Traiwit

**Comment by WilkAndy** [comment, marked as answer]
I'm not sure that your question makes sense, @Traiwit .   "Implicit" and "explicit" usually refer to the timestepping method, but if you're doing steadystate only then there is no time.  To put it differently - to achieve steadystate, you'd have to time-step to infinity, which would take an explicit solver an infinite amount of time.

**Reply by Traiwit** [reply]
Thank you for your answer Andy, I just realised that I didn't think of the fundamental of FEM. 
I'm now kinda dig myself into a hole with this mining simulation i'm trying to build.

I might work on quasi-static instead (you suggested a while ago), where after the elastic steady-state is reached > change the marterial property of 1 block > run steady-state simulation again > repeat.

I'm not sure if MOOSE can do that, I know it's quite simple with for-loop. This might worth posting in a new discussion.

**Reply by WilkAndy** [reply]
If you do that, you probably want to run it in a single simulation.  Make the model `transient` but have no time-derivative (velocity, acceleration, etc) Kernels.  Then each timestep will be solving the steady-state problem, ie, you'll be doing a quasi-static simulation.  The "time" in these simulations isn't a real time, but is simply to help you divide up the mining excavations into stages.  Eg, set dt=1, and in the first timestep you mine stage 1, in the next timestep you mine stage 2, etc.  So the notion of "time" essentially allows you to do your "for loop" idea.

An associated advantage is that you can, if you wish, shove in some time-derivative kernels, and solve the dynamic problem (you'll probably need damping), which can help with convergence when you've got complicated plasticity, and you can use an explicit method if you like.

**Reply by Traiwit** [reply]
Hi Andy, thank you for your reply.

Do you have (or can think of) any example for this quasi-static simulation? I tried to make the input file transient, but the error said 'i'm the a wrong state' for some reason, maybe because I do not have 'time' term anywhere in the input file.

again, thank you!

**Comment by Traiwit** [comment]
@WilkAndy I read your coal-mining example, I see your problem works well with quasi-static as you set the ecavation as a function of time and location

```
[./excav_sideways]
    type = ParsedFunction
    vars = 'end_t ymin ymax  minval maxval slope'
    vals = '17.0   0    1000.0 1E-9 1 60'
    # excavation face at ymin+(ymax-ymin)*min(t/end_t,1)
    # slope is the distance over which the modulus reduces from maxval to minval
    value = 'if(y<ymin+(ymax-ymin)*min(t/end_t,1),minval,if(y<ymin+(ymax-ymin)*min(t/end_t,1)+slope,minval+(maxval-minval)*(y-(ymin+(ymax-ymin)*min(t/end_t,1)))/slope,maxval))'
  [../]
```

However, is it possible to do quasi-static with mining-step plan associate with block ID for example
t = 1 // excavate block 1 2
t = 2 // excavate block 3 4 5
t = 3 // excavate block 6 7 and backfill block 2
etc ...

with my very little knowledge of MOOSE, i doubt MOOSE can do what i mentioned above in a single input file.
I could do that by running each steady-state simulation using the solution from previous step.

what do you think?

Best regards,
Traiwit

**Reply by WilkAndy** [reply]
Yes, this is quite easy.   Just define an appropriate `ParsedFunction` .  It'll look horrible, eg
```
value = 'if(t<=1, if(y<2000, 1, 0), if(t <=2 , if(y<5000 & y >= 2000, 1, 0)))'
```

**Reply by Traiwit** [reply]
Hi Andy, very interesting that MOOSE can do that, and I agree it will be very messy if the excavation blocks are 3D in the middle of the geometry. I will give it a go and get back to you.

I just hope that it can be done via blockID instead of defining the dimensions
for example: at t=1 E of block1 = 0 & E for all the other blocks remain the same as the previous step

Kind regards,
Trai

**Reply by WilkAndy** [reply]
That will be no problem, since you can define your ElasticityTensor block-by-block and deactivate it using it's own `elasticity_tensor_prefactor`.   Real mining probably happens more like a [MovingPlanarFront](https://mooseframework.inl.gov/source/functions/MovingPlanarFront.html) which allows you to specify a start position, an end position, a mining velocity, and an `active_length` (which could be your void, for instance).

**Reply by Traiwit** [reply]
Hi @WilkAndy thank you again for your suggestion, I will give it a go and probs get back to you in the afternoon.

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
@WilkAndy it works, but if my input has like 400+ blocks i might need a script to write an input file.

Now, I fill need to work with backfilling and then I can move on to plasticity.

again, thank you very much for your support!

Below is the example case, 0 = geology block, 1 2 3 = excavating blocks

```

[Functions]
[./E_excav1]
  type = ParsedFunction
  value = 'if(t>=1,10e-4,1)'
[../]
[./E_excav2]
  type = ParsedFunction
  value = 'if(t>=2,10e-4,1)'
[../]
[./E_excav3]
  type = ParsedFunction
  value = 'if(t>=3,10e-4,1)'
[../]

[./density_excav1]
  type = ParsedFunction
  value = 'if(t>=1,1.225,2700)' # air density or rock density
[../]
[./density_excav2]
  type = ParsedFunction
  value = 'if(t>=2,1.225,2700)' # air density or rock density
[../]
[./density_excav3]
  type = ParsedFunction
  value = 'if(t>=3,1.225,2700)' # air density or rock density
[../]
[]

[Materials]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

[./elasticity_tensor_0_unmined]
  type = ComputeIsotropicElasticityTensor
  block = '0'
  youngs_modulus = 20e9
  poissons_ratio = 0.25
[../]
[./elasticity_tensor_1]
  type = ComputeIsotropicElasticityTensor
  block = '1'
  youngs_modulus = 20e9
  poissons_ratio = 0.25
  elasticity_tensor_prefactor = E_excav1
[../]
[./elasticity_tensor_2]
  type = ComputeIsotropicElasticityTensor
  block = '2'
  youngs_modulus = 20e9
  poissons_ratio = 0.25
  elasticity_tensor_prefactor = E_excav2
[../]
[./elasticity_tensor_3]
  type = ComputeIsotropicElasticityTensor
  block = '3'
  youngs_modulus = 20e9
  poissons_ratio = 0.25
  elasticity_tensor_prefactor = E_excav3
[../]

[./density_0_unmined]
  type = GenericConstantMaterial
  block = '0'
  prop_names = density
  prop_values = 2700
[../]
[./density_1]
  type = GenericFunctionMaterial
  block = '1'
  prop_names = density
  prop_values = density_excav1
[../]
[./density_2]
  type = GenericFunctionMaterial
  block = '2'
  prop_names = density
  prop_values = density_excav2
[../]
[./density_3]
  type = GenericFunctionMaterial
  block = '3'
  prop_names = density
  prop_values = density_excav3
[../]
[]

```
